{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp readuniprot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# readuniprot\n",
    "\n",
    "> Parse Uniprot/Swissprot text format files (.dat), and Uniref XML files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from nbdev import show_doc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "\n",
    "import collections\n",
    "import csv\n",
    "import itertools\n",
    "import random\n",
    "import re\n",
    "\n",
    "from proteinscan import utils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code to parse Uniprot/Swissprot .dat files (text format protein database):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def iterDat(datFPath) :\n",
    "    \"\"\"\n",
    "    Iterator yielding a list of lines for each entry in the .dat file.\n",
    "    \"\"\"\n",
    "    with utils.openGzipOrText(datFPath) as f :\n",
    "        for (isEnd,it) in itertools.groupby(f, lambda ln : ln.strip()=='//') :\n",
    "            if not isEnd :\n",
    "                yield(list(it))\n",
    "\n",
    "def datEntryLnsWithCode(datEntry,code) :\n",
    "    \"Returns a list of the lines in a .dat file entry with the given code.\"\n",
    "    res = []\n",
    "    for ln in datEntry :\n",
    "        ln = ln.strip()\n",
    "        if ln.startswith(code+' ') :\n",
    "            res.append(ln[len(code):].lstrip())\n",
    "    return res\n",
    "\n",
    "def datEntryPrimaryAC(datEntry) :\n",
    "    \"Returns the primary accession number from a .det file entry.\"\n",
    "    acLns = datEntryLnsWithCode(datEntry,'AC')\n",
    "    return acLns[0].split(';')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "entries = list(iterDat('uniprotTest.dat.gz'))\n",
    "assert (len(entries)==185\n",
    "        and len(entries[2])==61\n",
    "        and entries[2][0]=='ID   002R_IIV3               Reviewed;         458 AA.\\n'\n",
    "        and entries[2][10]=='OX   NCBI_TaxID=345201;\\n'\n",
    "        and entries[2][-1]=='     QSIDRYFCSL DSNYNSEDED FEYDSDSEDD DSDSEDDC\\n'\n",
    "        and datEntryLnsWithCode(entries[2],'DT') ==\n",
    "            ['16-JUN-2009, integrated into UniProtKB/Swiss-Prot.',\n",
    "             '11-JUL-2006, sequence version 1.',\n",
    "             '02-JUN-2021, entry version 28.']\n",
    "        and datEntryPrimaryAC(entries[2])=='Q197F8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def scanDat(datFPath, fn, returnFull=False, **kwargs) :\n",
    "    \"\"\"\n",
    "    Scan a function across each entry in a .dat file, accumulating the results.\n",
    "    If the function returns None, the entry is ignored; otherwise the result\n",
    "    is added to the returned results list. Accepts additional keyword arguments,\n",
    "    which are passed on to the function to be scanned.\n",
    "\n",
    "    If returnFull is True, also returns a list of the corresponding full entries.\n",
    "    \"\"\"\n",
    "    res = []\n",
    "    if returnFull : fullRes = []\n",
    "    for datEntry in iterDat(datFPath) :\n",
    "        item = fn(datEntry, **kwargs)\n",
    "        if item is not None :\n",
    "            res.append(item)\n",
    "            if returnFull :\n",
    "                fullRes.append(datEntry)\n",
    "    return (res,fullRes) if returnFull else res\n",
    "\n",
    "def allPrimaryACsInDat(datFPath) :\n",
    "    \"Returns a list of all primary accession numbers in a .dat file.\"\n",
    "    return scanDat(datFPath,datEntryPrimaryAC)\n",
    "\n",
    "def datEntryPE(datEntry) :\n",
    "    \"Returns the level of evidence for a protein's existence (1-5, lower is stronger).\"\n",
    "    peLns = datEntryLnsWithCode(datEntry,'PE')\n",
    "    return int(peLns[0][0])\n",
    "\n",
    "def datEntryName(datEntry) :\n",
    "    \"Returns a descriptive name from a .dat file entry.\"\n",
    "    deLns = datEntryLnsWithCode(datEntry,'DE')\n",
    "    name = deLns[0].split('=')[1].rstrip(';')\n",
    "    flagSet = set()\n",
    "    for deLn in deLns[1:] :\n",
    "        if deLn.startswith('Flags') :\n",
    "            flagSet.update(flag.rstrip(';') for flag in deLn.split()[1:])\n",
    "    if len(flagSet) > 0 :\n",
    "        name += ' (' + '; '.join(sorted(flagSet)) +')'\n",
    "    return name\n",
    "\n",
    "def datEntrySeq(datEntry) :\n",
    "    \"Returns the amino acid sequence from a .dat file entry as a single string.\"\n",
    "    seqL = []\n",
    "    for ln in reversed(datEntry) :\n",
    "        ln = ln.strip()\n",
    "        if ln.startswith('SQ ') :\n",
    "            break\n",
    "        seqL.append(ln.replace(' ',''))\n",
    "    return ''.join(reversed(seqL))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acs = allPrimaryACsInDat('uniprotTest.dat.gz')\n",
    "pes = scanDat('uniprotTest.dat.gz',datEntryPE)\n",
    "names = scanDat('uniprotTest.dat.gz',datEntryName)\n",
    "seqs = scanDat('uniprotTest.dat.gz',datEntrySeq)\n",
    "assert (len(acs)==185\n",
    "        and acs[:5]==['Q6GZX4', 'Q6GZX3', 'Q197F8', 'Q197F7', 'Q6GZX2']\n",
    "        and acs[-5:]==['P0C9G3', 'P0C9G4', 'P0C9G1', 'P18559', 'P0C9G6']\n",
    "        and len(pes)==185\n",
    "        and pes[:5]==[4, 4, 4, 4, 3]\n",
    "        and pes[-5:]==[3, 3, 3, 3, 3]\n",
    "        and len(names)==185\n",
    "        and names[:3]==['Putative transcription factor 001R',\n",
    "                      'Uncharacterized protein 002L', 'Uncharacterized protein 002R']\n",
    "        and names[-3:]==['Protein MGF 110-1L', 'Protein MGF 110-2L (Precursor)',\n",
    "                         'Protein MGF 110-2L (Precursor)']\n",
    "        and len(seqs)==185\n",
    "        and len(seqs[0])==256\n",
    "        and seqs[0]=='MAFSAEDVLKEYDRRRRMEALLLSLYYPNDRKLLDYKEWSPPRVQVECPKAPVEWNNPPSEKGLIVGH'\n",
    "            +'FSGIKYKGEKAQASEVDVNKMCCWVSKFKDAMRRYQGIQTCKIPGKVLSDLDAKIKAYNLTVEGVEGFVRYSRVTK'\n",
    "            +'QHVAAFLKELRHSKQYENVNLIHYILTDKRVDIQHLEKDLVKDFKALVESAHRMRQGHMINVKYILYQLLKKHGHG'\n",
    "            +'PDGPDILTVKTGSKGVLYDDSFRKIYTDLGWKFTPL'\n",
    "        and len(seqs[-1])==113\n",
    "        and seqs[-1]=='MGFFSYLGLVLVGLASLTSLASLANLQDFSTDNPLEEELRCWCQYVKNCRFCWACQDGFCKNKVLK'\n",
    "            +'NMPSVQEHSYPMEHCMIHRQCKYVRDGPIFQVECMMQTCDAIHLLNA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def datEntryGOLines(datEntry) :\n",
    "    \"Returns the lines with GO terms from the database x-ref section of a .dat file entry.\"\n",
    "    drLns = datEntryLnsWithCode(datEntry,'DR')\n",
    "    return [drLn for drLn in drLns if drLn.startswith('GO;')]\n",
    "\n",
    "def datEntryGOTermPresent(datEntry,goTerm) :\n",
    "    \"Returns 'pos' or 'neg' accordingly as a .dat file entry includes the given GO term.\"\n",
    "    if any(goTerm in goLn for goLn in datEntryGOLines(datEntry)) :\n",
    "        return 'pos'\n",
    "    return 'neg'\n",
    "\n",
    "def datEntryIsAtpBinding(datEntry) :\n",
    "    \"Returns 'pos' or 'neg accordingly as a .dat file entry has the ATP binding GO term.\"\n",
    "    return datEntryGOTermPresent(datEntry,'GO:0005524;')\n",
    "def datEntryIsGtpBinding(datEntry) :\n",
    "    \"Returns 'pos' or 'neg accordingly as a .dat file entry has the GTP binding GO term.\"\n",
    "    return datEntryGOTermPresent(datEntry,'GO:0005525;')\n",
    "def datEntryIsMetalBinding(datEntry) :\n",
    "    \"Returns 'pos' or 'neg accordingly as a .dat file entry has the metal binding GO term.\"\n",
    "    return datEntryGOTermPresent(datEntry,'GO:0046872;')\n",
    "\n",
    "def datEntryKWs(datEntry) :\n",
    "    \"Returns a list of the keywords from the KW section of a .dat file entry.\"\n",
    "    res = []\n",
    "    for kwLn in datEntryLnsWithCode(datEntry,'KW') :\n",
    "        for kw in kwLn.split(';') :\n",
    "            kw = kw.strip().rstrip('.').strip().lower()\n",
    "            if kw != '' :\n",
    "                res.append(kw)\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "goLns = scanDat('uniprotTest.dat.gz',datEntryGOLines)\n",
    "kws = scanDat('uniprotTest.dat.gz',datEntryKWs)\n",
    "assert (goLns[0]==['GO; GO:0046782; P:regulation of viral transcription; IEA:InterPro.']\n",
    "        and goLns[1]==['GO; GO:0033644; C:host cell membrane; IEA:UniProtKB-SubCell.',\n",
    "                       'GO; GO:0016021; C:integral component of membrane; IEA:UniProtKB-KW.']\n",
    "        and goLns[-1]==[]\n",
    "        and datEntryGOTermPresent(entries[0],'GO:0046782')=='pos'\n",
    "        and datEntryGOTermPresent(entries[1],'GO:0033644')=='pos'\n",
    "        and datEntryGOTermPresent(entries[1],'GO:0016021')=='pos'\n",
    "        and datEntryGOTermPresent(entries[1],'GO:0046782')=='neg'\n",
    "        and datEntryGOTermPresent(entries[-1],'GO:0046782')=='neg'\n",
    "        and kws[0]==['activator','reference proteome','transcription',\n",
    "                     'transcription regulation']\n",
    "        and kws[1]==['host membrane','membrane','reference proteome',\n",
    "                     'transmembrane','transmembrane helix']\n",
    "        and kws[-1]==['early protein', 'signal'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "aaLetters = 'ACDEFGHIKLMNPQRSTVWY'\n",
    "aaLettersSet = set(aaLetters)\n",
    "\n",
    "def listIfStr(lOrStr) :\n",
    "    return [lOrStr] if isinstance(lOrStr,str) else lOrStr\n",
    "\n",
    "def filterDatEntry(datEntry, restrictTo20AA=True,\n",
    "            minLen=50, maxLen=400, maxPE=None,\n",
    "            requireInSpecies='', elimKWs=[], requireKWs=[],\n",
    "            requireInName='', excludeStrs=[]) :\n",
    "    \"\"\"\n",
    "    Filters a .dat file entry for inclusion in a LM or classification data set.\n",
    "    The filtering is based on the keyword arguments.\n",
    "\n",
    "    Returns (primary_accession_num,sequence) if the entry passes the filter, else None.\n",
    "    \"\"\"\n",
    "    seq = datEntrySeq(datEntry)\n",
    "    kws = datEntryKWs(datEntry)\n",
    "    elimKWs = listIfStr(elimKWs)\n",
    "    requireKWs = listIfStr(requireKWs)\n",
    "    excludeStrs = listIfStr(excludeStrs)\n",
    "    if ((restrictTo20AA and any(c not in aaLettersSet for c in seq))\n",
    "        or any(elimKW in kws for elimKW in elimKWs)\n",
    "        or any(requireKW not in kws for requireKW in requireKWs)\n",
    "        or (minLen is not None and len(seq)<minLen)\n",
    "        or (maxLen is not None and len(seq)>maxLen)\n",
    "        or (maxPE is not None and datEntryPE(datEntry)>maxPE)\n",
    "        or requireInSpecies not in datEntryLnsWithCode(datEntry,'OS')[0].lower()) :\n",
    "        return None\n",
    "    lName = datEntryName(datEntry).lower()\n",
    "    if ((requireInName not in lName)\n",
    "        or any (excludeStr in lName for excludeStr in excludeStrs)) :\n",
    "        return None\n",
    "    return (datEntryPrimaryAC(datEntry),seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (len(scanDat('uniprotTest.dat.gz',filterDatEntry))==166\n",
    "        and len(scanDat('uniprotTest.dat.gz',filterDatEntry,maxPE=3))==50\n",
    "        and len(scanDat('uniprotTest.dat.gz',filterDatEntry,maxPE=3,\n",
    "                        requireKWs=['transmembrane']))==15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code to parse UniRef files (precalculated clusters of similar protein sequences):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### UniRef XML format\n",
    "\n",
    "UniRef files contain precomputed clusters of similar protein sequences,\n",
    "at different thresholds of sequence identity (100, 90, 50 - these are percentages).\n",
    "UniRef XML files are named, `unirefN.xml.gz` where N is the identity threshold (for example `uniref50.xml.gz`).\n",
    "These files define a set of clusters, each of which contains a set of member sequences.\n",
    "\n",
    "Each cluster has two special member sequences (these may be the same) - a **seed** sequence, which is simply the longest\n",
    "sequence in the cluster, and a **representative** sequence, which is chosen to be the most \"informative\" sequence\n",
    "in the cluster using a number of criteria including quality of information available, length, etc.\n",
    "Cluster identifiers are constructed from the prefix `UniRef`,\n",
    "followed by the sequence identity threshold (`100`, `90`, or `50`),\n",
    "followed by underscore and the UniProt accession number or UniParc id of the chosen **representative** sequence for the cluster\n",
    "(for example, `UniRef50_A0A5A9P0L4` or `UniRef50_UPI0016133188`).\n",
    "\n",
    "<!-- A sequence in a cluster will always be at least N% identical to the seed sequence of that cluster,\n",
    "where N is the identity threshold (100, 90, or 50). In the case of UniRef90 and UniRef50,\n",
    "a sequence will also always be at least 11 AA long and have at least 80% overlap with the seed sequence;\n",
    "these requirements aren't applied to UniRef100.\n",
    " -->\n",
    "\n",
    "In more detail, the XML format consists of a series of `entry` elements. Each `entry` defines a cluster,\n",
    "and consists of one `representativeMember` element and a series of additional `member` elements,\n",
    "with each member identifying one sequence in the cluster.\n",
    "Each `entry` also has some attributes and elements describing the cluster as a whole,\n",
    "including an `id` attribute containing the cluster identifier constructed as described above,\n",
    "a `name` element with a descriptive name,\n",
    "and some `property` elements with different `type` attributes, including `type=\"member_count\"`\n",
    "(number of sequences in the cluster), `type=\"common taxon\"`, `type=\"common taxon ID\"`,\n",
    "and `type=\"GO Molecular Function\"` (I assume these last elements contain\n",
    "GO terms that are common to all members of the cluster).\n",
    "\n",
    "The `representativeMember` and `member` elements contain a `dbReference` element\n",
    "identifying the member sequence. This element includes:\n",
    "- Attributes `type` specifying a database (for example `UniParc ID` or `UniProtKB ID`) and `id` within that database (for example `UPI000FFEDAD7` or `A0A410P257_9BACT`)\n",
    "- If the sequence is in UniProt, one or more `property` elements with `type=\"UniProtKB accession\"` giving UniProt accession numbers for the sequence (for example `A0A410P257`) - if more than one, these will all be alternates for the same sequence in UniProt.\n",
    "- If the top level `id` attribute is from UniProt, a `property` element with `type=\"UniParc ID\"` giving the UniParc id for the sequence.\n",
    "- For UniRef90 and UniRef50, a `property` element with `type=\"UniRef100 ID\"` and `value=` the UniRef100 cluster id for the sequence.\n",
    "- For UniRef50, a `property` element with `type=\"UniRef90 ID\"` and `value=` the UniRef90 cluster id for the sequence.\n",
    "- Some `property` elements with information about the sequence, including `type=\"length\"`, `type=\"protein name\"`, `type=\"source organism\"`, `type=\"NCBI taxonomy\"`\n",
    "\n",
    "The `representativeMember` elements also contain a `sequence` element giving the actual AA sequence.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export \n",
    "\n",
    "def parseClustersFromUniref(unirefFPath,clusterFPath,encoding='ISO-8859-1') :\n",
    "    \"\"\"\n",
    "    Reads a UniRef XML file and generates a file just giving the cluster info.\n",
    "    Each line of the file is a space-separated list of primary accession numbers\n",
    "    from UniProt, which have been grouped into a cluster in the UniRef file.\n",
    "    Only keeps the entries in UniRef that have a UniProt accession number.\n",
    "    \"\"\"\n",
    "    nEntries = nAccNos = 0\n",
    "    with utils.openGzipOrText(unirefFPath,encoding) as f :\n",
    "        with open(clusterFPath,'w') as outF :\n",
    "            for i,ln in enumerate(f) :\n",
    "                if (i%100000000)==0 :  # print every 100-millionth line\n",
    "                    print(i,ln.strip())\n",
    "                if '<entry' in ln :  # start a cluster\n",
    "                    memCount = 0\n",
    "                    expectedMemCount = None\n",
    "                    entryAccNos = []\n",
    "                    nEntries += 1\n",
    "                elif '</entry' in ln :  # end a cluster\n",
    "                    if memCount != expectedMemCount :\n",
    "                        print('member count mismatch',memCount,expectedMemCount)\n",
    "                    nAccNos += len(entryAccNos)\n",
    "                    if len(entryAccNos) >= 1 :\n",
    "                        outF.write(\" \".join(entryAccNos)+'\\n')\n",
    "                elif '<member' in ln or '<representativeMember' in ln :  # start a sequence member\n",
    "                    memAccNos = []\n",
    "                elif '</member' in ln or '</representativeMember' in ln :  # end a sequence member\n",
    "                    entryAccNos.extend(memAccNos)\n",
    "                    memCount += 1\n",
    "                elif '<property' in ln :  # parse a property line\n",
    "                    m = re.search(r'<property.*type=\"(.*?)\".*value=\"(.*?)\"',ln)\n",
    "                    if m is not None :\n",
    "                        t,v = m.groups()\n",
    "                        if t.lower()=='member count' :  # member count from entry element\n",
    "                            expectedMemCount = int(v.strip())\n",
    "                        elif t.lower()=='uniprotkb accession' : # accession number from member element\n",
    "                            memAccNos.append(v.strip())\n",
    "    print(nEntries,'entries',nAccNos,'accNos')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 <?xml version=\"1.0\" encoding=\"ISO-8859-1\" ?>\n",
      "8 entries 1278 accNos\n"
     ]
    }
   ],
   "source": [
    "parseClustersFromUniref('unirefTest.xml.gz','unirefTestClusters.txt')\n",
    "unirefL = [ln.split() for ln in open('unirefTestClusters.txt').read().splitlines()]\n",
    "assert ([len(accNoL) for accNoL in unirefL] == [1, 1, 1232, 22, 1, 21]\n",
    "        and unirefL[0]==['A0A5A9P0L4']\n",
    "        and unirefL[2][:5]==['Q8WZ42', 'A6NKB1', 'E7EQE6', 'E7ET18', 'K7ENY1']\n",
    "        and unirefL[2][-5:]==['A0A2D4NV39', 'A0A2D4GXH3', 'A0A6P7QTE9', 'A0A6P7QKP2', 'A0A6P7QQP5']\n",
    "        and unirefL[-1]==['K7G060', 'M7AMC7', 'G3URQ7', 'G1NB19', 'A0A091K6H9', 'A0A7K4ZDL0',\n",
    "                          'A0A7L0GLZ9', 'A0A7L0TQB1', 'A0A7L1IDD4', 'G3UQ76', 'G3UR09', 'A0A7K5D0T2',\n",
    "                          'A0A7K9EPM6', 'A0A091EPU3', 'A0A091EPV7', 'A0A091ESE9', 'A0A7K7BU71',\n",
    "                          'A0A7L1PTK0', 'A0A7L1UQH9', 'A0A7K7BUC8', 'A0A7L4KN96']\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# experiment\n",
    "assert fastai is not None"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
